src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,48 +3,166 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+
+    private int size;
+    private Node<T> first;
+    private Node<T> last;
+
+    public MyLinkedList() {
+        this.size = 0;
+        first = new Node<>(null, null, null);
+        last = new Node<>(null, null, null);
+    }
+
+    private void backElement(T tail) {
+        Node<T> last = this.last;
+        Node<T> node = new Node<>(last, tail, (Node) null);
+        if (last.element == null) {
+            this.first = node;
+        } else {
+            last.next = node;
+        }
+        this.last = node;
+        this.size++;
+    }
+
+    private void frontElement(T head) {
+        Node<T> first = this.first;
+        Node<T> node = new Node<>((Node) null, head, first);
+        if (first.element == null) {
+            this.last = node;
+        } else {
+            first.prev = node;
+        }
+        this.first = node;
+        this.size++;
+    }
+
     @Override
     public void add(T value) {
-
+        if (first.element == null) {
+            frontElement(value);
+        } else {
+            backElement(value);
+        }
     }
 
     @Override
     public void add(T value, int index) {
-
+        if (index < 0 || index > size) {
+            throw new IndexOutOfBoundsException();
+        }
+        if (index == size) {
+            add(value);
+        } else {
+            Node<T> prev = first;
+            for (int i = 0; i < size; i++) {
+                if (index == i) {
+                    Node<T> node = new Node<>(prev, value, prev.next);
+                    prev = node;
+                    if (prev == null) {
+                        this.first = node;
+                    } else {
+                        prev.next = node;
+                    }
+                    size++;
+                    break;
+                }
+                prev = prev.next;
+            }
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndexException(index);
+        Node<T> node = first;
+        for (int i = 0; i < index; ++i) {
+            node = node.next;
+        }
+        return node.element;
     }
 
     @Override
     public void set(T value, int index) {
-
+        checkIndexException(index);
+        Node<T> node = first;
```

Comment:
this code is the same that code in the method get)
better move it in the separate method